File: 백준/Platinum/14860. GCD 곱/GCD 곱.py
import sys 
print = lambda *args, sep=" ", end="\n": sys.stdout.write(sep.join(map(str, args)) + end)
input = lambda: sys.stdin.readline().rstrip('\r\n')

MOD = 1_000_000_007

def main():
    n, m = map(int, input().split())
    if n > m: n, m = m, n
    size = n//3+1
    isp = bytearray([1])*size
    isp[0] = 0
    limit = int(n**0.5)
    for i in range(1, limit//3+1):
        if isp[i]:
            p = 3*i+1|1
            d, s, j = p*2, p*p, p*(p+4-2*(i&1))
            isp[s//3::d] = bytearray([0])*((size-1-s//3)//d+1)
            isp[j//3::d] = bytearray([0])*((size-1-j//3)//d+1)
    primes = [2, 3]+[3*i+1|1 for i in range(1, size) if isp[i]]
    ans = 1
    for p in primes:
        pi, ex = p, 0
        while pi <= n:
            ex = (ex+(n//pi)*(m//pi))
            pi *= p
        ans = ans*pow(p, ex%(MOD-1), MOD)%MOD
    print(ans)
    return
if __name__ == '__main__':
    main()

File: 백준/Platinum/14860. GCD 곱/GCD 곱.cc
#include<bits/stdc++.h>
using namespace std;
using ll = long long;

const ll MOD = 1'000'000'007;

ll pw(ll n, ll r, ll MOD){
    n %= MOD;
    ll ret = 1;
    while(r){
        if(r&1) ret = ret*n%MOD;
        n = n*n%MOD;
        r >>= 1;
    }
    return ret;
}

int main(){
    cin.tie(0); cout.tie(0); ios::sync_with_stdio(false);
    int n, m; cin >> n >> m;
    if(n > m) swap(n, m);
    int size = n/3+1;
    vector<bool> isp(size, 1);
    isp[0] = 0;
    for(int i = 1; i*i <= size; i++){
        if(isp[i]){
            int p = 3*i+1|1;
            int d = p*2;
            int s = p*p;
            int j = p*(p+4-2*(i&1));
            for(int k = s/3; k < size; k += d) isp[k] = 0;
            for(int k = j/3; k < size; k += d) isp[k] = 0;
        }
    }
    vector<int> primes = {2, 3};
    for(int i = 1; i < size; i++) if(isp[i]) primes.push_back(3*i+1|1);
    ll ans = 1;
    for(auto p : primes){
        ll pi = p, ex = 0;
        while(pi <= n){
            ex = (ex+(n/pi)*(m/pi));
            pi *= p;
        }
        ans = ans*pw(p, ex%(MOD-1), MOD)%MOD;
    }
    cout << ans;
    return 0;
}